include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -5748,17 +5761,53 @@ namespace internal
             evaluator.template hessians<1, true, false>(out, out);
           if (dim > 2)
             evaluator.template hessians<2, true, false>(out, out);
+        }
+      if (not dyadic_coefficients)
+        {
+          const unsigned int shift_coefficient =
+            inverse_coefficients.size() > dofs_per_component ?
+              dofs_per_component :
+              0;
+
+          for (unsigned int q = 0; q < dofs_per_component; ++q)
+            for (unsigned int d = 0; d < n_desired_components; ++d)
+              out_array[q + d * dofs_per_component] *=
+                inv_coefficient[q + d * shift_coefficient];
+        }
+      else
+        {
+          const unsigned int n_coeff_components =
+            n_desired_components * n_desired_components;
 
           for (unsigned int q = 0; q < dofs_per_component; ++q)
-            out[q] *= inv_coefficient[q];
+            {
+              std::vector<Number> tmp(n_desired_components);
+              for (unsigned int d = 0; d < n_desired_components; ++d)
+                tmp[d] = out_array[q + d * dofs_per_component];
 
+              // This is basically a matrix-vector-product
+              for (unsigned int d1 = 0; d1 < n_desired_components; ++d1)
+                {
+                  Number sum(0.);
+                  for (unsigned int d2 = 0; d2 < n_desired_components; ++d2)
+                    {
+                      sum +=
+                        inverse_coefficients[q * n_coeff_components +
+                                             d1 * n_desired_components + d2] *
+                        tmp[d2];
+                    }
+                  out_array[q + d1 * dofs_per_component] = sum;
+                }
```

Comment:
May I ask you to move this code (matrix-vector product) into a helper function? I would also suggest to template the function according to the number of components. At the caller side, I would write:
```cpp
if(n_coeff_components==1)
  vmult<1>(...); // scalar
if(n_coeff_components==dim)
  vmult<dim>(...); // vector Laplace
else
  vmult<-1>(...);
```

What do you think?